synapse/http/replicationagent.py

```diff
@@ -0,0 +1,170 @@
+class ReplicationEndpointFactory:
+    """Connect to a given TCP socket"""
+
+    def __init__(
+        self,
+        reactor: ISynapseReactor,
+        context_factory: IPolicyForHTTPS,
+    ) -> None:
+        self.reactor = reactor
+        self.context_factory = context_factory
+
+    def endpointForURI(self, uri: URI) -> IStreamClientEndpoint:
+        """
+        This part of the factory decides what kind of endpoint is being connected to.
+
+        Args:
+            uri: The pre-parsed URI object containing all the uri data
+
+        Returns: The correct client endpoint object
+        """
+        if uri.scheme in (b"http", b"https"):
+            endpoint = HostnameEndpoint(self.reactor, uri.host, uri.port)
+            if uri.scheme == b"https":
+                endpoint = wrapClientTLS(
+                    self.context_factory.creatorForNetloc(uri.host, uri.port), endpoint
+                )
+            return endpoint
+        else:
+            raise SchemeNotSupported(f"Unsupported scheme: {uri.scheme!r}")
+
+
+@implementer(IAgent)
+class ReplicationAgent(_AgentBase):
+    """
+    This Agent is solely for the purposes of connecting to Synapse replication
+    endpoints, and can handle https and http connections. Appropriate comments are
+    copied from Twisted's Agent Class.
+
+    Attributes:
+        _endpointFactory: The IAgentEndpointFactory which will
+            be used to create endpoints for outgoing TCP connections.
+    """
+
+    def __init__(
+        self,
+        reactor: ISynapseReactor,
+        contextFactory: Optional[IPolicyForHTTPS] = None,
+        connectTimeout: Optional[float] = None,
+        bindAddress: Optional[bytes] = None,
+        pool: Optional[HTTPConnectionPool] = None,
+    ):
+        """
+        Create a ReplicationAgent.
+
+        Args:
+            reactor: A reactor for this Agent to place outgoing connections.
+            contextFactory: A factory for TLS contexts, to control the
+                verification parameters of OpenSSL.  The default is to use a
+                BrowserLikePolicyForHTTPS, so unless you have special
+                requirements you can leave this as-is.
+            connectTimeout: The amount of time that this Agent will wait
+                for the peer to accept a connection.
+            bindAddress: The local address for client sockets to bind to.
+            pool: An HTTPConnectionPool instance, or None, in which
+                case a non-persistent HTTPConnectionPool instance will be
+                created.
+        """
+        if not IPolicyForHTTPS.providedBy(contextFactory):
+            logger.warning(
+                f"{contextFactory} was passed as the HTTPS policy for an "
+                "Agent, but it does not provide IPolicyForHTTPS.  Since Twisted 14.0, "
+                "you must pass a provider of IPolicyForHTTPS.",
+            )
+            contextFactory = _DeprecatedToCurrentPolicyForHTTPS(contextFactory)
```

Comment:
I don't think we need to copy this backwards compatibility code (unless we do this?)

Reply:
I think this might be safe to lose. We use the Synapse context factory which has everything that `IPolicyForHTTPS` requires(in this case the `creatorForNetloc()` (see [docs](https://docs.twisted.org/en/stable/api/twisted.web.iweb.IPolicyForHTTPS.html))

Reply:
Done in [144474d](https://github.com/matrix-org/synapse/pull/15470/commits/144474d5d0f3d5ae8b5f638c3108111127fe9cf9)